### PR TITLE
Make stringify_keys use transform_keys

### PR DESCRIPTION
**Pull request by styd**
Since Ruby 2.5, there is `Hash#transform_keys` that runs faster than the traditional way of transforming keys. From my benchmark, it's around 11-13% faster on 2.5 and above.
So, I'm thinking of turning `stringify_keys` to use `transform_keys` to take the performance benefit and I also make the polyfill for Ruby less than 2.5.

Here is my benchmark:
```ruby
require 'benchmark/ips'

def stringify_keys(other)
  hash = {}
  other.each do |key, value|
    hash[key.to_s] = value
  end
  hash
end

def stringify_keys2(other)
  other.transform_keys(&:to_s)
end

hash = {a: 'an a', b: 'a b', az: 'a and z', zzz: 'sleeping..'}

Benchmark.ips do |x|
  x.report('traditional') do
    stringify_keys(hash)
  end

  x.report('transform') do
    stringify_keys2(hash)
  end

  x.compare!
end
```
```
Warming up --------------------------------------
         traditional    85.231k i/100ms
           transform    93.393k i/100ms
Calculating -------------------------------------
         traditional      1.005M (± 2.3%) i/s -      5.029M in   5.004178s
           transform      1.132M (± 0.3%) i/s -      5.697M in   5.032428s

Comparison:
           transform:  1132060.4 i/s
         traditional:  1005440.7 i/s - 1.13x  slower
```

### CODE REVIEW

File: lib/rack/core_ext/hash.rb [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# frozen_string_literal: true
+
+# Hash has `transform_keys` since Ruby 2.5.
+# For Ruby < 2.5, we need to add the following
+
+module Rack
+  module HashExtensions
+    if Gem::Version.new(RUBY_VERSION) < Gem::Version.new(2.5)
```
Comment (rafaelfranca) [inline lib/rack/core_ext/hash.rb]: Do we need to check the version given we are checking if `{}.respond_to?(:transform_keys)`?

Comment (styd) [reply lib/rack/core_ext/hash.rb]: I was thinking of making it more explicit about what version of ruby will this be used for. But, I guess I already add it in the comment.

Comment (styd) [reply lib/rack/core_ext/hash.rb]: Ok, removed.

File: lib/rack/session/abstract/id.rb [hunk 2]
```diff
@@ -17,6 +17,17 @@ module Abstract
...
       class SessionHash
+        using Module.new {
+          refine Hash do
+            def transform_keys(&block)
+              hash = {}
+              each do |key, value|
+                hash[block.call(key)] = value
+              end
+              hash
+            end unless {}.respond_to?(:transform_keys)
```
Comment (rafaelfranca) [inline lib/rack/session/abstract/id.rb]: Should we move the `{}.respond_to?(:transform_keys)` to the `using` call to avoid refining on Ruby 2.6?

Comment (styd) [reply lib/rack/session/abstract/id.rb]: You're right. That seems more efficient.

**Comment by jkowens** [comment]
@styd Any idea why we're seeing the following failure in our Sinatra tests?

```
1) Error:
HelpersTest::TestSession#test_creates_a_new_session_when_none_provided_0:
NoMethodError: undefined method `transform_keys' for #<Rack::Session::Abstract::SessionHash:0x0000000003cebb90>
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/session/abstract/id.rb:177:in `stringify_keys'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/session/abstract/id.rb:113:in `update'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/session/abstract/id.rb:282:in `prepare_session'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/session/abstract/id.rb:241:in `context'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/session/abstract/id.rb:236:in `call'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/null_logger.rb:11:in `call'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/head.rb:14:in `call'
    /home/travis/build/sinatra/sinatra/lib/sinatra/base.rb:194:in `call'
    /home/travis/build/sinatra/sinatra/lib/sinatra/base.rb:1950:in `call'
    /home/travis/build/sinatra/sinatra/lib/sinatra/base.rb:1502:in `block in call'
    /home/travis/build/sinatra/sinatra/lib/sinatra/base.rb:1729:in `synchronize'
    /home/travis/build/sinatra/sinatra/lib/sinatra/base.rb:1502:in `call'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/lint.rb:51:in `_call'
    /home/travis/.rvm/gems/ruby-2.4.9/bundler/gems/rack-fb2668fec7b9/lib/rack/lint.rb:39:in `call'
    /home/travis/.rvm/gems/ruby-2.4.9/gems/rack-test-1.1.0/lib/rack/mock_session.rb:29:in `request'
    /home/travis/.rvm/gems/ruby-2.4.9/gems/rack-test-1.1.0/lib/rack/test.rb:266:in `process_request'
    /home/travis/.rvm/gems/ruby-2.4.9/gems/rack-test-1.1.0/lib/rack/test.rb:129:in `custom_request'
    /home/travis/.rvm/gems/ruby-2.4.9/gems/rack-test-1.1.0/lib/rack/test.rb:58:in `get'
    /home/travis/.rvm/gems/ruby-2.4.9/gems/rack-test-1.1.0/lib/rack/test.rb:203:in `follow_redirect!'
    /home/travis/build/sinatra/sinatra/test/helpers_test.rb:506:in `block (2 levels) in <class:HelpersTest>'
```

**Comment by styd** [comment]
@jkowens, maybe my new PR can fix it.

**Comment by jkowens** [comment]
Great, thanks!

**Pull request merged**